src/dense.jl

```diff
@@ -1,193 +1,209 @@
-# you would lose the opprotunity to detect the roots with each step.
-function next{O<:ExplicitODE,S<:DenseStepper}(s::Solver{O,S}, state::DenseState)
-
-
-    # m3: I'm not 100% sure what happens here.  I would implement it like so:
-    # Initialize in `start`: calculate next t1, y1 and also hold onto IC in t0,y0,
-    # set state.last_t=1
-    #
-    # in next have this loop
-    # for t in tspan[state.last_t:end]
-    #     if t>t1
-    #        make new t1, y1, move old t1, y1 into t0, y0
-    #     end
-    #     make dense output at t
-    #     find events
-    #     state.last_t += 1
-    #     return ((t, y), state)
-    # end
-
-    # pwl: @m3 this is basically what happens here:-), although I'm
-    # not using the index of tspan anywhere explicitly.
-
-    solver = s.stepper.solver
-
-    # these guys store the intermediate steps we make
-    s0, s1 = state.s0, state.s1
-    t0, t1 = s0.t, s1.t
-
-    # assuming the last output was done at state.last_tout set the
-    # t_goal to the next larger time from tspan.  Strong inequality
-    # below is crucial, otherwise we would be selecting the same step
-    # every time.
-    tspan  = s.options.tspan
-    t_goal = tspan[findfirst(t->(t>state.last_tout), tspan)]
-
-    # Keep computing new steps (i.e. new pairs (t0,t1)) until we reach
-    # t0 < t_goal <= t1, then we use interpolation to get the value at
-    # t_goal.  Unless points==:all, then we break the while loop after
-    # making the first step.
-    while t_goal > t1
-
-        # s1 stores the last succesfull step, the new step is stored
-        # in s0
-
-        if done(solver, state.solver_state)
-            warn("The iterator was exhausted before the dense output completed.")
-            # prevents calling done(..) twice
-            state.solver_done = true
-            # TODO: deepcopy?
-            # Return whatever we got as the last step
-            return ((s0.t,s0.y[:]),state)
-        else
-            # at this point s0 is updated with the new step, "s2" if you will
-            ((s0.t,s0.y[:]), state.solver_state) = next(solver, state.solver_state)
-        end
 
-        # swap s0 and s1
-        s0, s1 = s1, s0
-        # and times
-        t0, t1 = s0.t, s1.t
+"""
 
-        # update the state accordingly
-        state.s0, state.s1 = s0, s1
+TODO: rename `tout` to `tout` and drop the support for
+`points=:all` outside of the `odeXX`?  Maybe even
+`odeXX(;tout=[...])` would use dense output while `odeXX(;)`
+wouldn't.
 
-        # we haven't reached t_goal yet (t1<t_goal) but the option
-        # points==:all calls for an output at every successful
-        # step.
-        if s.options.points == :all && t1 < t_goal-eps(t1)
-            state.last_tout = t1
-            return ((t1,s1.y),state)
-        end
+"""
 
-        if s.options.stopevent(t1,s1.y)
-            break
-        end
+function onestep!(ode::ExplicitODE,
+                  stepper::DenseStepper,
+                  state::DenseState)
```

Comment:
Right, we are not restricted to `ExplicitODE` now!  `dstate` seems good.  Could you update this in your #9?